Lib/dbm/sqlite3.py

```diff
@@ -59,19 +59,25 @@ def __init__(self, path, /, *, flag, mode):
 
         # We use the URI format when opening the database.
         uri = _normalize_uri(path)
-        uri = f"{uri}?mode={flag}"
+        if flag == "ro":
+            # Add immutable=1 to allow read-only SQLite access even if wal/shm missing
+            uri = f"{uri}?mode={flag}&immutable=1"
+        else:
+            uri = f"{uri}?mode={flag}"
 
         try:
             self._cx = sqlite3.connect(uri, autocommit=True, uri=True)
         except sqlite3.Error as exc:
             raise error(str(exc))
 
+        self._readonly = (flag == "ro")
         # This is an optimization only; it's ok if it fails.
-        with suppress(sqlite3.OperationalError):
-            self._cx.execute("PRAGMA journal_mode = wal")
+        if not self._readonly:
```

Comment:
`self._readonly` is not used anywhere else, is it? You could just use `flag != "ro"` directly:
```suggestion
        if flag != "ro":
```